Bluetooth: GATT: Various optimizations

Vudentz opened the pull request:
This adds optimization for handle lookup into the database and persist the database hash in flash so it can be used to check if service changed needs to be emitted on reboot.

Vudentz committed:
Bluetooth: GATT: Skip ahead if handle is not within service handles

This reduces bt_gatt_foreach_attr complexity (O(n)) so it can skip
ahead when the start handle is not within its service handles.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -933,6 +933,15 @@ void bt_gatt_foreach_attr(u16_t start_handle, u16_t end_handle,
+		next = SYS_SLIST_PEEK_NEXT_CONTAINER(svc, node);
+		if (next) {
+			/* Skip ahead if start is not within service handles */
+			if (next->attrs[0].handle <= start_handle) {
```
Should this really be <= and not < ?

Vudentz replied:
It is the next service so equal still means is not in the current svc.

jhedberg replied:
right

carlescufi replied:
`next->attrs[0].handle` is the handle of the service definition of the next service? If so I don't get why you would skip ahead in that case

Vudentz replied:
Service, and its attributes, are registered in ascending order, so this is checking if the handle happens to be >= the following service which means is not withing the current service so it just skip iterating into the service handles. It is allowed to set a fixed handle, provided it is not in use, which is why we dont simple use handle + num_handles logic, though the logic is quite similar.

Vudentz committed:
Bluetooth: GATT: Fix iterating over requested handle range

If handle is over the range end return since handles are always placed
in ascending order.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/settings.c:
```diff
@@ -89,20 +89,15 @@ int bt_settings_decode_key(char *key, bt_addr_le_t *addr)
+	for (h = _bt_settings_start; h < _bt_settings_end; h++) {
+		if (!strcmp(argv[0], h->name)) {
+			argc--;
+			argv++;
```
I guess this change to the generic bt settings code should be split into its own commit with an explanation of what/why.

Vudentz replied:
Alright, it seems the top level (bt/) was restricted to just keys settings so this is changing to allow settings also only on top level.

Vudentz replied:
Actually my last comment is not quite right, the problem seems to be that we always expect an extra argument when in fact there could be settings that don't require one.

Vudentz replied:
It is now a separate patch.

Vudentz committed:
Bluetooth: Settings: Fix not allowing custom handlers without arguments

Custom handlers without any arguments don't since there is a check for
argc > 1, to fix this behavior removing the check.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Store database hash

This stores the database hash and check if it has changed on commit
skipping service changed if hash matches.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>
